SMB2 connections not closed due to SMBApiException

When multiple messages are written to a shared location concurrently using SMB2, the following error is thrown for some of the write attempts:
```
[2024-03-29 13:28:43,494] ERROR {VFSTransportSender} - [Service: sample] - cannot resolve replyFile repeatedly: Could not establish connection to "localhost". org.apache.commons.vfs2.FileSystemException: Could not establish connection to "localhost".
        at org.apache.commons.vfs2.provider.smb2.Smb2ClientWrapper.setupClient(Smb2ClientWrapper.java:117)
        at org.apache.commons.vfs2.provider.smb2.Smb2ClientWrapper.<init>(Smb2ClientWrapper.java:72)
        at org.apache.commons.vfs2.provider.smb2.Smb2WrapperClientFactory.create(Smb2WrapperClientFactory.java:36)
        at org.apache.commons.vfs2.provider.smb2.Smb2FileSystem.<init>(Smb2FileSystem.java:49)
        at org.apache.commons.vfs2.provider.smb2.Smb2FileProvider.doCreateFileSystem(Smb2FileProvider.java:58)
        at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.getFileSystem(AbstractOriginatingFileProvider.java:158)
        at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.findFile(AbstractOriginatingFileProvider.java:112)
        at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.findFile(AbstractOriginatingFileProvider.java:82)
        at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:789)
        ....
Caused by: com.hierynomus.mssmb2.SMBApiException: STATUS_REQUEST_NOT_ACCEPTED (0xc00000d0): Authentication failed for 'User-1' using com.hierynomus.smbj.auth.NtlmAuthenticator@3dc75b64
        at com.hierynomus.smbj.connection.SMBSessionBuilder.setupSession(SMBSessionBuilder.java:144)
        at com.hierynomus.smbj.connection.SMBSessionBuilder.setupSession(SMBSessionBuilder.java:142)
        at com.hierynomus.smbj.connection.SMBSessionBuilder.establish(SMBSessionBuilder.java:109)
        at com.hierynomus.smbj.connection.Connection.authenticate(Connection.java:202)
        at org.apache.commons.vfs2.provider.smb2.Smb2ClientWrapper.setupClient(Smb2ClientWrapper.java:100)
        ... 41 more
```

The connections for which the error is thrown remain open until they eventually timeout. This results in multiple open smb2 connections in the server.

--- a/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2ClientWrapper.java
+++ b/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2ClientWrapper.java
@@ -104,16 +104,7 @@ public class Smb2ClientWrapper extends SMBClient {
             if (LOG.isDebugEnabled()) {
                 LOG.debug("Error while creating connection to " + rootName.getHostName());
             }
-            try {
-                if (session != null) {
-                    session.close();
-                }
-                if (connection != null) {
-                    connection.close();
-                }
-            } catch (IOException ignored) {
-
-            }
+            close();
             throw new FileSystemException("vfs.provider.smb2/connect.error", rootName.getHostName(), e);
         }
     }
@@ -243,6 +234,33 @@ public class Smb2ClientWrapper extends SMBClient {
             diskShare.rmdir(path, true);
         } else {
             diskShare.rm(path);
+        }
+    }
+
+    public void close() {
+        try {
+            if (diskShare != null) {
+                diskShare.close();
+            }
+            if (session != null) {
+                session.close();
+            }
+            if (connection != null) {
+                connection.close(true);
+            }
+        } catch (Exception e) {
+            LOG.debug("Error while closing the connection.", e);
+        } finally {
+            if (connection != null) {
+                try {
+                    LOG.debug("Connection closed forcefully for : " + smbClient);
+                    connection.close(true);
+                } catch (IOException e) {
+                    //Ignore
+                }
+            }
+            // Close the client
+            super.close();
         }
     }
 }

--- a/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2FileSystem.java
+++ b/commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2FileSystem.java
@@ -95,10 +95,10 @@ public class Smb2FileSystem extends AbstractFileSystem {
      */
     public void putClient(SMBClient smbClient) {
         if (isClosed) {
-            smbClient.close();
+            ((Smb2ClientWrapper)smbClient).close();
         } else if (!client.compareAndSet(null, smbClient)) {
             // An idle client is already present so close the connection.
-            smbClient.close();
+            ((Smb2ClientWrapper)smbClient).close();
         }
     }
 }

